File: 2071_maximum_number_of_tasks_you_can_assign.py
from bisect import bisect_left
from sortedcontainers import SortedList

class Solution:
    def maxTaskAssign(self, tasks, workers, pills, strength):
        tasks.sort()
        workers.sort()

        def can_do(k):
            wl = SortedList(workers[-k:])
            p = pills
            for i in range(k - 1, -1, -1):
                if wl[-1] >= tasks[i]:
                    wl.pop()
                else:
                    if p == 0:
                        return False
                    idx = wl.bisect_left(tasks[i] - strength)
                    if idx == len(wl):
                        return False
                    wl.pop(idx)
                    p -= 1
            return True

        l, r = 0, min(len(tasks), len(workers))
        res = 0
        while l <= r:
            m = (l + r) // 2
            if can_do(m):
                res = m
                l = m + 1
            else:
                r = m - 1
        return res

File: 2071_maximum_number_of_tasks_you_can_assign.cpp
class Solution {
public:
    int maxTaskAssign(vector<int>& a, vector<int>& b, int p, int s) {
        sort(a.begin(), a.end()), sort(b.begin(), b.end());
        int l = 0, h = min(a.size(), b.size()), r = 0;
        while (l <= h) {
            int m = (l + h) / 2, x = p;
            multiset<int> z(b.end() - m, b.end());
            bool f = 1;
            for (int i = m - 1; i >= 0; --i) {
                auto it = prev(z.end());
                if (*it >= a[i]) z.erase(it);
                else if (x) {
                    it = z.lower_bound(a[i] - s);
                    if (it == z.end()) { f = 0; break; }
                    z.erase(it); --x;
                } else { f = 0; break; }
            }
            f ? r = m, l = m + 1 : h = m - 1;
        }
        return r;
    }
};
